File: taurus-protect-sdk-go/pkg/protect/mapper/blockchain.go
package mapper

import (
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/openapi"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// BlockchainFromDTO converts an OpenAPI TgvalidatordBlockchainEntity to a domain Blockchain.
func BlockchainFromDTO(dto *openapi.TgvalidatordBlockchainEntity) *model.Blockchain {
	if dto == nil {
		return nil
	}

	blockchain := &model.Blockchain{
		Symbol:           safeString(dto.Symbol),
		Name:             safeString(dto.Name),
		Network:          safeString(dto.Network),
		ChainID:          safeString(dto.ChainId),
		BlackholeAddress: safeString(dto.BlackholeAddress),
		Confirmations:    safeString(dto.Confirmations),
		BlockHeight:      safeString(dto.BlockHeight),
		IsLayer2Chain:    safeBool(dto.IsLayer2Chain),
		Layer1Network:    safeString(dto.Layer1Network),
	}

	// Convert base currency
	if dto.BaseCurrency != nil {
		blockchain.BaseCurrency = CurrencyFromDTO(dto.BaseCurrency)
	}

	// Convert EVM info
	if dto.EthInfo != nil {
		blockchain.EVMInfo = EVMBlockchainInfoFromDTO(dto.EthInfo)
	}

	// Convert DOT info
	if dto.DotInfo != nil {
		blockchain.DOTInfo = DOTBlockchainInfoFromDTO(dto.DotInfo)
	}

	// Convert XTZ info
	if dto.XtzInfo != nil {
		blockchain.XTZInfo = XTZBlockchainInfoFromDTO(dto.XtzInfo)
	}

	return blockchain
}

// BlockchainsFromDTO converts a slice of OpenAPI TgvalidatordBlockchainEntity to domain Blockchains.
func BlockchainsFromDTO(dtos []openapi.TgvalidatordBlockchainEntity) []*model.Blockchain {
	if dtos == nil {
		return nil
	}
	blockchains := make([]*model.Blockchain, len(dtos))
	for i := range dtos {
		blockchains[i] = BlockchainFromDTO(&dtos[i])
	}
	return blockchains
}

// EVMBlockchainInfoFromDTO converts an OpenAPI TgvalidatordEVMBlockchainInfo to a domain EVMBlockchainInfo.
func EVMBlockchainInfoFromDTO(dto *openapi.TgvalidatordEVMBlockchainInfo) *model.EVMBlockchainInfo {
	if dto == nil {
		return nil
	}

	return &model.EVMBlockchainInfo{
		ChainID: safeString(dto.ChainId),
	}
}

// DOTBlockchainInfoFromDTO converts an OpenAPI TgvalidatordDOTBlockchainInfo to a domain DOTBlockchainInfo.
func DOTBlockchainInfoFromDTO(dto *openapi.TgvalidatordDOTBlockchainInfo) *model.DOTBlockchainInfo {
	if dto == nil {
		return nil
	}

	info := &model.DOTBlockchainInfo{
		CurrentEra:     safeString(dto.CurrentEra),
		MaxNominations: safeString(dto.MaxNominations),
		ForkNumber:     safeString(dto.ForkNumber),
	}

	if dto.ForkMigratedAt != nil {
		info.ForkMigratedAt = dto.ForkMigratedAt
	}

	return info
}

// XTZBlockchainInfoFromDTO converts an OpenAPI TgvalidatordXTZBlockchainInfo to a domain XTZBlockchainInfo.
func XTZBlockchainInfoFromDTO(dto *openapi.TgvalidatordXTZBlockchainInfo) *model.XTZBlockchainInfo {
	if dto == nil {
		return nil
	}

	return &model.XTZBlockchainInfo{
		CurrentCycle: safeString(dto.CurrentCycle),
	}
}

File: taurus-protect-sdk-typescript/src/mappers/blockchain.ts
/**
 * Blockchain mapper functions for converting OpenAPI DTOs to domain models.
 */

import type {
  Blockchain,
  DOTBlockchainInfo,
  EVMBlockchainInfo,
  XTZBlockchainInfo,
} from '../models/blockchain';
import { currencyFromDto } from './currency';
import { safeBool, safeInt, safeMap, safeString } from './base';

/**
 * Maps a DOT blockchain info DTO to domain model.
 */
function dotBlockchainInfoFromDto(dto: unknown): DOTBlockchainInfo | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    ss58Format: safeInt(d.ss58Format ?? d.ss_58_format),
  };
}

/**
 * Maps an EVM blockchain info DTO to domain model.
 */
function evmBlockchainInfoFromDto(dto: unknown): EVMBlockchainInfo | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    chainId: safeString(d.chainId ?? d.chain_id),
  };
}

/**
 * Maps an XTZ blockchain info DTO to domain model.
 */
function xtzBlockchainInfoFromDto(dto: unknown): XTZBlockchainInfo | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    protocolHash: safeString(d.protocolHash ?? d.protocol_hash),
  };
}

/**
 * Maps a blockchain DTO to a Blockchain domain model.
 */
export function blockchainFromDto(dto: unknown): Blockchain | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    symbol: safeString(d.symbol),
    name: safeString(d.name),
    network: safeString(d.network),
    chainId: safeString(d.chainId ?? d.chain_id),
    confirmations: safeString(d.confirmations),
    blockHeight: safeString(d.blockHeight ?? d.block_height),
    blackholeAddress: safeString(d.blackholeAddress ?? d.blackhole_address),
    isLayer2Chain: safeBool(d.isLayer2Chain ?? d.is_layer2_chain),
    layer1Network: safeString(d.layer1Network ?? d.layer1_network),
    baseCurrency: currencyFromDto(d.baseCurrency ?? d.base_currency),
    dotInfo: dotBlockchainInfoFromDto(d.dotInfo ?? d.dot_info),
    ethInfo: evmBlockchainInfoFromDto(d.ethInfo ?? d.eth_info),
    xtzInfo: xtzBlockchainInfoFromDto(d.xtzInfo ?? d.xtz_info),
  };
}

/**
 * Maps an array of blockchain DTOs to Blockchain domain models.
 */
export function blockchainsFromDto(dtos: unknown[] | null | undefined): Blockchain[] {
  return safeMap(dtos, blockchainFromDto);
}
